build_tools/scripts/byo_llvm.sh

```diff
@@ -0,0 +1,175 @@
+#!/bin/bash
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+TD="$(cd $(dirname $0) && pwd)"
+REPO_ROOT="$(cd $TD/../.. && pwd)"
+LLVM_SOURCE_DIR="${LLVM_SOURCE_DIR:-${REPO_ROOT}/third_party/llvm-project}"
+LLVM_BUILD_DIR="${LLVM_BUILD_DIR:-${REPO_ROOT}/../iree-llvm-build}"
+LLVM_INSTALL_DIR="${LLVM_INSTALL_DIR:-${REPO_ROOT}/../iree-llvm-install}"
+
+command="$1"
+shift
+
+# Detect commands.
+has_ccache=false
+if (command -v ccache &> /dev/null); then
+  has_ccache=true
+fi
+has_clang=false
+if (command -v clang &> /dev/null) && (command -v clang++ &> /dev/null); then
+  has_clang=true
+fi
+has_lld=false
+if (command -v lld &> /dev/null); then
+  has_lld=true
+fi
+python3_command=""
+if (command -v python3 &> /dev/null); then
+  python3_command="python3"
+elif (command -v python &> /dev/null); then
+  python3_command="python"
+fi
+
+set -euo pipefail
+
+print_toolchain_config() {
+  if $has_ccache; then
+    echo -n "-DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache "
+  fi
+  if $has_clang; then
+    echo "-DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++"
+  fi
+}
+
+do_build_llvm() {
+  echo "*********************** BUILDING LLVM *********************************"
+  main_build_dir="${LLVM_BUILD_DIR}/llvm"
+  main_install_dir="${LLVM_INSTALL_DIR}/llvm"
+  targets_to_build="${LLVM_TARGETS_TO_BUILD:-X86}"
+  enable_projects="${LLVM_ENABLE_TARGETS:-clang;lld}"
+
+  cmake_options="-DLLVM_ENABLE_PROJECTS='${enable_projects}' -DLLVM_TARGETS_TO_BUILD='${targets_to_build}'"
+  cmake_options="${cmake_options} -DLLVM_BUILD_EXAMPLES=OFF"
+  cmake_options="${cmake_options} -DLLVM_INSTALL_UTILS=ON"
+  cmake_options="${cmake_options} -DCMAKE_BUILD_TYPE=Release"
+  cmake_options="${cmake_options} -DLLVM_ENABLE_ASSERTIONS=ON"
```

Comment:
We may need to make this configurable or a way for user to specify all non-essential flags.

Reply:
Added some comments about how this can evolve. I don't intend this to be the one script to rule them all for all possible uses but a starting point for people to adapt to their dev flow with this one remaining a thing we can verify on CI.